### PR TITLE
tests: update to latest

### PR DESCRIPTION
**Pull request by holiman**
This PR updates the tests, which presumably contains some shanghai tests.

### CODE REVIEW

**Comment by holiman** [comment]
This has many failures like `TestState/EIPTests/stEOF/stEIP3540/CREATE2_EOF1Invalid_FromEOF.json/Shanghai` -- this implies that it is generated with `EOF` in `Shanghai`. Which is not the case. 
I've made it so those are now skipped.

**Comment by holiman** [comment]
The block tests fail. I think this is the place that is "wrong" 
```
--- FAIL: TestBlockchain (0.03s)
    --- FAIL: TestBlockchain/EIPTests/bc4895-withdrawals/shanghaiWithoutWithdrawalsRLP.json (0.01s)
        block_test.go:52: test without snapshotter failed: block (index 0) insertion should have failed due to: RLP_ExpectedAsList
        block_test.go:55: test with snapshotter failed: block (index 0) insertion should have failed due to: RLP_ExpectedAsList
    --- FAIL: TestBlockchain/ValidBlocks/bcExample/shanghaiExample.json (0.00s)
        block_test.go:52: test without snapshotter failed: genesis block hash doesn't match test: computed=151934ad9b65, test=18db39e19931
        block_test.go:55: test with snapshotter failed: genesis block hash doesn't match test: computed=151934ad9b65, test=18db39e19931
```
For the first failure, I think this is the place that is "wrong" -- it only checks the length, not if it is nil. The block is actually IMO "correct", it's just that our representation of the block body is missing a field, something that we can easily address internally. So I'm not sure I agree. If someone hands us an incomplete block (which we can fix up locally), is it correct to reject it? 



```golang
	// Fail if Shanghai not enabled and len(withdrawals) is non-zero.
	withdrawals := block.Withdrawals()
	if len(withdrawals) > 0 && !p.config.IsShanghai(block.Time()) {
		return nil, nil, 0, fmt.Errorf("withdrawals before shanghai")
	}
```

As for the second error, I have no idea what that is all about.

**Comment by holiman** [comment]
Ah, so the second error seems to be, that the `genesis` in the test has this field 
```
            "withdrawalsRoot" : "0x1dcc4de8dec75d7aab85b567b6ccd41ad312451b948a7413f0a142fd40d49347"
```
However, `0x1dcc4de8dec75d7aab85b567b6ccd41ad312451b948a7413f0a142fd40d49347` -- this is the "empty list hash" which is used e.g. for the `ommerHash`. But the withdrawalsroot (if empty) should use the empty root hash, `56e81f171bcc55a6ff8345e692c0f86e5b48e01b996cadc001622fb5e363b421`. 

So I'd say the test is wrong there, cc @winsvega

**Comment by winsvega** [comment]
Yes, the test is wrong.

**Comment by holiman** [comment]
This fixes the last (?) failure
```diff
diff --git a/core/state_processor.go b/core/state_processor.go
index 163ea0a020..df5dfaf516 100644
--- a/core/state_processor.go
+++ b/core/state_processor.go
@@ -90,6 +90,8 @@ func (p *StateProcessor) Process(block *types.Block, statedb *state.StateDB, cfg
 	withdrawals := block.Withdrawals()
 	if len(withdrawals) > 0 && !p.config.IsShanghai(block.Time()) {
 		return nil, nil, 0, fmt.Errorf("withdrawals before shanghai")
+	} else if withdrawals == nil && p.config.IsShanghai(block.Time()) {
+		return nil, nil, 0, fmt.Errorf("nil withdrawals post-shanghai")
 	}
 	// Finalize the block, applying any consensus engine specific extras (e.g. block rewards)
 	p.engine.Finalize(p.bc, header, statedb, block.Transactions(), block.Uncles(), withdrawals)
```
One thing that is a bit bad with how it is _now_, is that with the check as is, we might not actually store an "empty list" withdrawals-body in the database, but totally missing the withdrawals. 

SO if we want to keep the block-check as is, we should also make sure to "populate" the block with the empty list, not just leave it `nil`. 
cc @fjl

**Comment by holiman** [comment]
Added this to triage, mainly to discuss 

> One thing that is a bit bad with how it is now, is that with the check as is, we might fail to store an "empty list" withdrawals-body in the database, and instead store one which is missing the withdrawals.

**Comment by fjl** [comment]
I have added the check, but it is misplaced in `StateProcessor`. This check should be done in `BlockValidator`.

**Review by fjl — APPROVED** [review]

File: tests/state_test.go [hunk 1]
```diff
@@ -56,11 +56,11 @@ func TestState(t *testing.T) {
...
 	st.skipLoad(`^stStaticCall/static_Call1MB`)
 
-	// Not yet supported TODO
-	st.skipLoad(`^stEIP3540/`)
-	st.skipLoad(`^stEIP3860/`)
-
 	// Broken tests:
+	//
+	// The stEOF tests are generated with EOF as part of Shanghai, which
+	// is erroenous. Therefore, these tests are skipped.
```
Comment (holiman) [inline tests/state_test.go:62]:
```suggestion
	// is erroneous. Therefore, these tests are skipped.
```

**Comment by holiman** [comment]
```
--- FAIL: TestStateProcessorErrors (0.02s)
    state_processor_test.go:373: test 0:
        have "missing withdrawals in block body"
        want "could not apply tx 0 [0x832b54a6c3359474a9f504b1003b2cc1b6fcaa18e4ef369eb45b5d40dad6378f]: max initcode size exceeded: code size 49153 limit 49152"
    state_processor_test.go:373: test 1:
        have "missing withdrawals in block body"
        want "could not apply tx 0 [0x39b7436cb432d3662a25626474282c5c4c1a213326fd87e4e18a91477bae98b2]: intrinsic gas too low: have 54299, want 54300"
FAIL
coverage: 69.8% of statements
FAIL	github.com/ethereum/go-ethereum/core	33.496s

```

**Pull request merged**